### PR TITLE
Res 1414 maxup

### PR DESCRIPTION
**Pull request by lucasosouza**
Maxup implementation, refactored as a mixin. 
Includes two versions: 
- MaxupStandard: the version which showed improved accuracy in experiments with dense networks in 100 and 1000 classes.
- MaxupPerSample: version that matches the method description in the paper. Experimental results are slightly lower than benchmark. Since the paper code is not available, we can't tell if the actual results shown in the paper follow the Standard or the PerSample implementation.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        :param data: input to the training function, as specified by dataloader
+        :param target: target to be matched by model, as specified by dataloader
+        :param async_gpu: define whether or not to use
+                          asynchronous GPU copies when the memory is pinned
+        """
+
+        if len(data.shape) < 5:
+            raise ValueError("Define replicas_per_sample > 1")
+
+        target = target.to(self.device, non_blocking=async_gpu)
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py:55]: It looks like it would be simpler to move this optimization to the callers in `ComplexLoss. _train_model`
https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/mixins/complex_loss.py#L71

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py [hunk 2]
```diff
@@ -335,6 +335,7 @@ def run_single_instance(config):
     config["workers"] = 4
     config["log_level"] = "INFO"
     config["reuse_actors"] = False
+    config["dist_port"] = get_free_port()
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:338]: This is for the run_single_instance only, not for the distributed. The get free port function from Ray has issues in the single instance case, if you break an experiment midway it will try to reuse the same port and fail.

File: nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py (hunk 1 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py:55]: The ComplexLoss is reused by other mixins, including Knowledge Distillation. In the Knowledge Distillation case, I don't need the target in some cases -  hence I get extra memory if I don't move target to the GPU, which is the reason I left this step to be done inside the calculate_batch_loss function.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 3]
```diff
@@ -156,6 +156,9 @@ def setup_experiment(self, config):
             - sample_transform: Transform acting on the training samples. To be used
                                 additively after default transform or auto-augment.
             - target_transform: Transform acting on the training targets.
+            - replicas_per_sample: Number of replicas to create per sample in the batch.
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py:159]: I thought about it, but couldn't find a good way to do it. `create_train_dataloader` is called in the middle of a very long init function in ImageNet experiment. If I let it be created and then override it after, the dataloaders will be instantiated twice. I could replace the entire `create_train_dataloader` function, passing a different function as argument. But I will still need access to the replicas per sample argument, and would have to change the call to `create_train_dataloader` anyway. Any ideas?

[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py:159]: You could use the same technique @mrcslws did for the loss_function. Create a real `create_train_dataloader` method that calls the private function `_create_train_dataloader`. This way we can keep the `replicas_per_sample` as part of the mixin that uses it.

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:338]: Got it.

File: nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py (hunk 1 above)
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py:55]: Makes sense.

Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/mixins/maxup.py:55]: In that case I think you should use the `async_gpu` local variable instead of forcing it to be true when calling `calculate_batch_loss`. See https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/mixins/complex_loss.py#L79

File: nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py:338]: You may need to change this line as well.
https://github.com/numenta/nupic.research/blob/f896cbd4e2b718c88059309c166cf9713d2dc237/nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py#L371

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py:159]: That is a good point. I am still thinking through the best way to work with mixins, and that is a good possibility, but in this case it still implicates similar changes in the code. I will merge as it is for now just to move this. I am currently working on something else that will require changing the dataloader, then if the changes are significant I will refactor everything out of ImagenetExperiment.
